src/yarn-exporter/src/yarn_exporter.py

```diff
@@ -16,128 +16,116 @@
+def request_with_histogram(url, histogram, *args, **kwargs):
+    with histogram.time():
+        return requests.get(url, *args, **kwargs)
 
 
-class YarnMetricCollector(YarnCollector):
-    api_path = '/ws/v1/cluster/metrics'
+class YarnCollector(object):
+    def __init__(self, yarn_url):
+        self.metric_url = urllib.parse.urljoin(yarn_url, "/ws/v1/cluster/metrics")
 
     def collect(self):
-        response = requests.get(self.metric_url, allow_redirects=True)
+        response = request_with_histogram(self.metric_url, cluster_metrics_histogram,
+                allow_redirects=True)
         response.raise_for_status()
-        metric = response.json()['clusterMetrics']
-
+        metric = response.json()["clusterMetrics"]
 
-        total_gpu_num = YarnMetric('total_gpu_num', YarnMetric.COUNTER, 
-                                    'The total number of GPUs of cluster',['cluster']).create_metric()
-        total_gpu_num.add_metric([self.cluster_name], metric['totalGPUs'])
+        total_gpu_num = gen_total_gpu_count()
+        total_gpu_num.add_metric([], metric["totalGPUs"])
         yield total_gpu_num
 
-        gpus_used = YarnMetric('gpus_used', YarnMetric.COUNTER, 
-                                    'The number of allocated GPUs',['cluster']).create_metric()
-        gpus_used.add_metric([self.cluster_name], metric['allocatedGPUs'])
+        gpus_used = gen_used_gpu_count()
+        gpus_used.add_metric([], metric["allocatedGPUs"])
         yield gpus_used
 
-        nodes_all = YarnMetric('nodes_all', YarnMetric.GAUGE,
-                               'The total number of nodes', ['cluster']).create_metric()
-        nodes_all.add_metric([self.cluster_name], metric['totalNodes'])
+        nodes_all = gen_total_gpu_count()
+        nodes_all.add_metric([], metric["totalNodes"])
         yield nodes_all
 
-        nodes_active = YarnMetric('nodes_active', YarnMetric.GAUGE,
-                                  'The number of active nodes', ['cluster']).create_metric()
-        nodes_active.add_metric([self.cluster_name], metric['activeNodes'])
+        nodes_active = gen_active_node_count()
+        nodes_active.add_metric([], metric["activeNodes"])
         yield nodes_active
 
-        # nodes_lost = YarnMetric('nodes_lost', YarnMetric.GAUGE,
-        #                         'The number of lost nodes', ['cluster']).create_metric()
-        # nodes_lost.add_metric([self.cluster_name], metric['lostNodes'])
-        # yield nodes_lost
 
-        # nodes_unhealthy = YarnMetric('nodes_unhealthy', YarnMetric.GAUGE,
-        #                              'The number of unhealthy nodes', ['cluster']).create_metric()
-        # nodes_unhealthy.add_metric([self.cluster_name], metric['unhealthyNodes'])
-        # yield nodes_unhealthy
+class HealthResource(Resource):
+    def render_GET(self, request):
+        request.setHeader("Content-Type", "text/html; charset=utf-8")
+        return "<html>Ok</html>".encode("utf-8")
```

Comment:
It's a python3 code, why need encode

Reply:
Twisted requires application returning byte, so encode is needed.